libgadget/forcetree.c

```diff
@@ -268,73 +268,85 @@ modify_internal_node(int parent, int subnode, int p_toplace, const ForceTree tb)
+                    p_toplace, P[p_toplace].Pos[0], P[p_toplace].Pos[1], P[p_toplace].Pos[2], P[p_toplace].Type, P[p_toplace].ID,
+                    oldsuns[0], P[oldsuns[0]].Pos[0], P[oldsuns[0]].Pos[1], P[oldsuns[0]].Pos[2], P[oldsuns[0]].Type, P[oldsuns[0]].ID,
+                    oldsuns[1], P[oldsuns[1]].Pos[0], P[oldsuns[1]].Pos[1], P[oldsuns[1]].Pos[2], P[oldsuns[1]].Type, P[oldsuns[1]].ID
+                );
+                return 1;
+            }
+            struct NODE *nfreep = &tb.Nodes[nprnt->u.s.suns[i]];
+            /* We create a new leaf node.*/
+            init_internal_node(nfreep, nprnt, i);
+            /*Set father of new node*/
+            nfreep->father = parent;
         }
-        struct NODE *nfreep = &tb.Nodes[nprnt->u.s.suns[i]];
-        /* We create a new leaf node.*/
-        init_internal_node(nfreep, nprnt, i);
-        /*Set father of new node*/
-        nfreep->father = parent;
-    }
-    /*Initialize the remaining entries to empty*/
-    for(i=8; i<NMAXCHILD;i++)
-        nprnt->u.s.suns[i] = -1;
-
-    for(i=0; i < NMAXCHILD; i++) {
-        /* Re-attach each particle to the appropriate new leaf.
-         * Notice that since we have NMAXCHILD slots on each child and NMAXCHILD particles,
-         * we will always have a free slot. */
-        int subnode = get_subnode(nprnt, oldsuns[i]);
+        /*Initialize the remaining entries to empty*/
+        for(i=8; i<NMAXCHILD;i++)
+            nprnt->u.s.suns[i] = -1;
+
+        for(i=0; i < NMAXCHILD; i++) {
+            /* Re-attach each particle to the appropriate new leaf.
+            * Notice that since we have NMAXCHILD slots on each child and NMAXCHILD particles,
+            * we will always have a free slot. */
+            int subnode = get_subnode(nprnt, oldsuns[i]);
+            int child = nprnt->u.s.suns[subnode];
+            struct NODE * nchild = &tb.Nodes[child];
+            modify_internal_node(child, nchild->u.s.noccupied, oldsuns[i], tb);
+            nchild->u.s.noccupied++;
+        }
+        /* Now try again to add the new particle*/
+        int subnode = get_subnode(nprnt, p_toplace);
         int child = nprnt->u.s.suns[subnode];
         struct NODE * nchild = &tb.Nodes[child];
-        modify_internal_node(child, nchild->u.s.noccupied, oldsuns[i], tb);
-        nchild->u.s.noccupied++;
-    }
-    /* Now try again to add the new particle*/
-    int subnode = get_subnode(nprnt, p_toplace);
-    int child = nprnt->u.s.suns[subnode];
-    struct NODE * nchild = &tb.Nodes[child];
-    if(nchild->u.s.noccupied < NMAXCHILD) {
-        modify_internal_node(child, nchild->u.s.noccupied, p_toplace, tb);
-        nchild->u.s.noccupied++;
-    }
-    else {
-        int ret = create_new_node_layer(child, p_toplace, tb, nnext, nc);
-        if(ret > 0)
-            return ret;
-    }
+        if(nchild->u.s.noccupied < NMAXCHILD) {
```

Comment:
I don't know if you want to mess with it more -- the current code does what it says. but
This logic is somewhat duplicated with line 480, which makes me wonder if the assertion on creating a new layer of nodes was only required to ease the recursion.

A simpler invariance is to assert that after a function call particle at to_place is put into the subtree starting at first_parent, and the new starting point of the insertion is returned.

Reply:
I see what you mean, but 480 is not necessarily a duplicate, because different threads may be walking the same subtree. The loop is useful because it takes place before the parent node is marked and unlocked, so it can happen without atomic accesses.

Reply:
Loop at 422? Yes I think we talked about that loop -- it indeed happens before locking. 

But line 480 we already have the lock on the tree branch rooting at `this`, so I think merging the logic from 480 to 486 into the function shouldn't break threading?

Reply:
Probably not - but I think I'd prefer to leave it in there, if that's ok. It doesn't hurt, and the code is already a bit confusing.